### ISSUE TITLE
What is the form of the output?

### ISSUE THREAD

**Issue by papb** [floydpink/lzwCompress.js#147]
Hello! I noticed that this module returns an array of numbers.

* What can I expect of the sizes of each number? The largest I got was 303 on a small example, and the smallest was 34. Is there a guaranteed lower bound? And upper bound?

* What can I expect of the length of this array?

**Comment by floydpink** [comment]
@papb - It looks like you might have confused the compression implemented in this library for something else. This library is useful for applying [the popular Lempel–Ziv–Welch](https://en.wikipedia.org/wiki/Lempel%E2%80%93Ziv%E2%80%93Welch) lossless compression algorithm on large JSON objects, so that it can be persisted into NoSQL databases or into the IndexedDB on the browser (for offline use).

As mentioned on the README, the output of the `pack()` call cannot be consumed directly - but if you persist this "packed" output instead of the original JSON into a NoSQL database, you should see the difference in the space utilization.

When you want to consume the "packed", large JSON object - you will then retrieve it from your persistence layer and the call `unpack()` from the library.

Does that help?

**Comment by papb** [comment]
Hello, thanks for the fast reply! I already knew what you said, sorry for not being clear on the question. I don't have any knowledge on how the LZW algorithm works - what I know is that it is a compression algorithm. The fact that the output of this algorithm consists of an array of numbers was something I didn't know, and learned by running this module. Now I want to know what can I expect about the bounds in these numbers and the array length, in order to think better on the persistence layer that I will use. For example, if I have a way to persist an array of numbers from 0 to 1023, can I use it? Or is it possible that this algorithm generates a number above 1023 at some point?

**Comment by floydpink** [comment]
I don't think we can definitively say what the upper bound is, for the numbers in the "compressed" array - and the reason it is like that is because the numbers will depend on the specific input passed into [the core LZW compress function](https://github.com/floydpink/lzwCompress.js/blob/main/lzwCompress.js#L119-L151) (which takes a random string and converts it into an array).

At the core (as can be seen on the function), the LZW algorithm starts with the ASCII character set loaded into a dictionary/map. And during the processing of a given input, it will start accumulating string segments that are repeated within.

Here is an example input of `"awesome awesome"`, that is being processed by this core `compress` function along with the data structure within the function visualized as it is processed: [Link to JavaScript Visualizer in PythonTutor.com](http://pythontutor.com/visualize.html#code=const%20compress%20%3D%20function%20%28uncompressed%29%20%7B%0A%20%20%20%20%20%20%20%20%20%20if%20%28typeof%20%28uncompressed%29%20!%3D%3D%20'string'%29%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20uncompressed%3B%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20let%20i%3B%0A%20%20%20%20%20%20%20%20%20%20const%20dictionary%20%3D%20Object.create%28null%29%3B%0A%20%20%20%20%20%20%20%20%20%20let%20c,%0A%20%20%20%20%20%20%20%20%20%20%20%20wc,%0A%20%20%20%20%20%20%20%20%20%20%20%20w%20%3D%20''%3B%0A%20%20%20%20%20%20%20%20%20%20const%20result%20%3D%20%5B%5D%3B%0A%20%20%20%20%20%20%20%20%20%20let%20dictSize%20%3D%20256%3B%0A%20%20%20%20%20%20%20%20%20%20for%20%28i%20%3D%200%3B%20i%20%3C%20256%3B%20i%20%2B%3D%201%29%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20dictionary%5BString.fromCharCode%28i%29%5D%20%3D%20i%3B%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20for%20%28i%20%3D%200%3B%20i%20%3C%20uncompressed.length%3B%20i%20%2B%3D%201%29%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20c%20%3D%20uncompressed.charAt%28i%29%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20wc%20%3D%20w%20%2B%20c%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20if%20%28dictionary%5Bwc%5D%29%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20w%20%3D%20wc%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%20else%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20if%20%28dictionary%5Bw%5D%20%3D%3D%3D%20undefined%29%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20return%20uncompressed%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20result.push%28dictionary%5Bw%5D%29%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20dictionary%5Bwc%5D%20%3D%20dictSize%2B%2B%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20w%20%3D%20String%28c%29%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20if%20%28w%20!%3D%3D%20''%29%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20result.push%28dictionary%5Bw%5D%29%3B%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20return%20result%3B%0A%20%20%20%20%20%20%20%20%7D%3B%0A%20%20%20%20%20%20%20%20%0Aconst%20result%20%3D%20compress%28'awesome%20awesome'%29%3B&cumulative=false&curInstr=0&heapPrimitives=nevernest&mode=display&origin=opt-frontend.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false)

As you can see, for this input - the numbers go all the way up to `260` - hopefully this gives some insight into what you are dealing with.
